HW2 Мельчук А.Б.

Тема “Создание признакового пространства”

Продолжим обработку данных с Твиттера. 

1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
	Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
	Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
	Исключим стоп-слова с помощью stop_words='english'.
	Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().




In [1]:
import warnings
warnings.filterwarnings("ignore")

In [2]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
import numpy as np
import pandas as pd

In [5]:
combine_df = pd.read_pickle('/content/combine_df.pkl')
combine_df.head()

id  ...                                   tweet_lemmatized
0   1  ...  [father, dysfunctional, selfish, drag, kid, dy...
1   2  ...  [thanks, lyft, credit, use, cause, offer, whee...
2   3  ...                                  [bihday, majesty]
3   4  ...                      [model, love, take, time, ur]
4   5  ...                  [factsguide, society, motivation]

[5 rows x 7 columns]

In [6]:
def tokenizer(text):
    return text

In [7]:
count_vectorizer = CountVectorizer(tokenizer=tokenizer, stop_words='english', lowercase=False, max_df=0.9,max_features = 1000)

# Создаем the Bag-of-Words модель
count_bag_of_words = count_vectorizer.fit_transform(combine_df['tweet_stemmed'].tolist())

# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer.get_feature_names()
pd.DataFrame(count_bag_of_words.toarray(), columns = feature_names)

abl  absolut  accept  account  act  ...  york  young  youtub  yr  yummi
0        0        0       0        0    0  ...     0      0       0   0      0
1        0        0       0        0    0  ...     0      0       0   0      0
2        0        0       0        0    0  ...     0      0       0   0      0
3        0        0       0        0    0  ...     0      0       0   0      0
4        0        0       0        0    0  ...     0      0       0   0      0
...    ...      ...     ...      ...  ...  ...   ...    ...     ...  ..    ...
49154    0        0       0        0    0  ...     0      0       0   0      0
49155    0        0       0        0    0  ...     0      0       0   0      0
49156    0        0       0        0    0  ...     0      0       0   0      0
49157    0        0       0        0    0  ...     0      0       0   0      0
49158    0        0       0        0    0  ...     0      0       0   0      0

[49159 rows x 1000 columns]

In [8]:
count_vectorizer_lem = CountVectorizer(tokenizer=tokenizer, stop_words='english',
                                   lowercase=False, max_df=0.9,max_features = 1000)

# Создаем the Bag-of-Words модель
count_bag_of_words_lem = count_vectorizer_lem.fit_transform(combine_df['tweet_lemmatized'].tolist())

# Отобразим Bag-of-Words модель как DataFrame
feature_names = count_vectorizer_lem.get_feature_names()
pd.DataFrame(count_bag_of_words_lem.toarray(), columns = feature_names)

able  absolutely  account  act  action  ...  young  youth  youtube  yr  yummy
0         0           0        0    0       0  ...      0      0        0   0      0
1         0           0        0    0       0  ...      0      0        0   0      0
2         0           0        0    0       0  ...      0      0        0   0      0
3         0           0        0    0       0  ...      0      0        0   0      0
4         0           0        0    0       0  ...      0      0        0   0      0
...     ...         ...      ...  ...     ...  ...    ...    ...      ...  ..    ...
49154     0           0        0    0       0  ...      0      0        0   0      0
49155     0           0        0    0       0  ...      0      0        0   0      0
49156     0           0        0    0       0  ...      0      0        0   0      0
49157     0           0        0    0       0  ...      0      0        0   0      0
49158     0           0        0    0       0  ...      0      0        0   0      0

[49159 rows x 1000 columns]

2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно.
	Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
	Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.
	Исключим стоп-слова с помощью stop_words='english'.
	Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().


In [9]:
tfidf_vectorizer = TfidfVectorizer(tokenizer=tokenizer, stop_words='english', lowercase=False, max_df=0.9,max_features = 1000)

# Создаем the Bag-of-Words модель
bag_of_words_tfidf = tfidf_vectorizer.fit_transform(combine_df['tweet_stemmed'].tolist())

# Отобразим Bag-of-Words модель как DataFrame
feature_names = tfidf_vectorizer.get_feature_names()
pd.DataFrame(bag_of_words_tfidf.toarray(), columns = feature_names)

abl  absolut  accept  account  act  ...  york  young  youtub   yr  yummi
0      0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
1      0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
2      0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
3      0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
4      0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
...    ...      ...     ...      ...  ...  ...   ...    ...     ...  ...    ...
49154  0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
49155  0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
49156  0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
49157  0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
49158  0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0

[49159 rows x 1000 columns]

In [12]:
tfidf_vectorizer_lem = TfidfVectorizer(tokenizer=tokenizer, stop_words='english',
                                   lowercase=False, max_df=0.9,max_features = 1000)

# Создаем the Bag-of-Words модель
bag_of_words_tfidf_lem = tfidf_vectorizer_lem.fit_transform(combine_df['tweet_lemmatized'].tolist())

# Отобразим Bag-of-Words модель как DataFrame
feature_names = tfidf_vectorizer_lem.get_feature_names()
pd.DataFrame(bag_of_words_tfidf_lem.toarray(), columns = feature_names)

able  absolutely  account  act  ...  youth  youtube   yr  yummy
0       0.0         0.0      0.0  0.0  ...    0.0      0.0  0.0    0.0
1       0.0         0.0      0.0  0.0  ...    0.0      0.0  0.0    0.0
2       0.0         0.0      0.0  0.0  ...    0.0      0.0  0.0    0.0
3       0.0         0.0      0.0  0.0  ...    0.0      0.0  0.0    0.0
4       0.0         0.0      0.0  0.0  ...    0.0      0.0  0.0    0.0
...     ...         ...      ...  ...  ...    ...      ...  ...    ...
49154   0.0         0.0      0.0  0.0  ...    0.0      0.0  0.0    0.0
49155   0.0         0.0      0.0  0.0  ...    0.0      0.0  0.0    0.0
49156   0.0         0.0      0.0  0.0  ...    0.0      0.0  0.0    0.0
49157   0.0         0.0      0.0  0.0  ...    0.0      0.0  0.0    0.0
49158   0.0         0.0      0.0  0.0  ...    0.0      0.0  0.0    0.0

[49159 rows x 1000 columns]

3. Натренируем gensim.models.Word2Vec модель на наших данных.
●	Тренировать будем на токенизированных твитах combine_df['tweet_token']
●	Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34.
●	Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.





In [13]:
import gensim.downloader as api

In [14]:
from gensim.models import Word2Vec, FastText

In [16]:
modelW2V = Word2Vec(sentences=combine_df['tweet_token'], size=200, window=5, min_count=2,
                    sg = 1, hs = 0, negative = 10, workers= 32, seed = 34)

4. Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".



In [17]:
modelW2V.most_similar(positive=['dinner'])

[('lunch', 0.869331419467926),
 ('bbq', 0.8609201908111572),
 ('drinks', 0.8564198017120361),
 ('chilled', 0.8543386459350586),
 ('bestie', 0.8481425046920776),
 ('brunch', 0.8469065427780151),
 ('picnic', 0.8419790267944336),
 ('familytime', 0.8419721126556396),
 ('lunchtime', 0.8413270115852356),
 ('cocktails', 0.8374963402748108)]

In [18]:
modelW2V.most_similar(positive=['trump'])

[('republican', 0.7664228081703186),
 ('obama', 0.737536609172821),
 ('fuhered', 0.7335864305496216),
 ('clinton', 0.7324715852737427),
 ('paladino', 0.7286496758460999),
 ('remarks', 0.7282741069793701),
 ('carl', 0.7276935577392578),
 ('racist', 0.7235125303268433),
 ('ally', 0.7226601839065552),
 ('blm', 0.7224089503288269)]

5. Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова).
Давайте проверим векторное представление любого слова из нашего корпуса, например "food".




In [19]:
modelW2V['food']

array([-0.0902303 , -0.34479475,  0.12639378, -0.28791177,  0.11492249,
       -0.05910985,  0.01896186,  0.2540192 ,  0.3294191 ,  0.280945  ,
        0.09672456,  0.06750061,  0.28285277, -0.425002  , -0.0898826 ,
        0.00145701, -0.09214064,  0.2050724 , -0.07698023, -0.19454016,
        0.13532843,  0.08428239, -0.02598125, -0.06878676, -0.02903615,
        0.04485585,  0.37732524,  0.1728817 ,  0.17690088,  0.19496076,
       -0.3319308 , -0.04447849,  0.4453062 , -0.09381836, -0.2198872 ,
        0.06971607, -0.37299564, -0.03281387, -0.21171075, -0.18381323,
        0.2573174 ,  0.14149097,  0.03910683, -0.41157305, -0.08387399,
        0.15480645, -0.03053956, -0.04075143,  0.52482545,  0.14861442,
        0.14372018, -0.36999846,  0.06562082,  0.19156525, -0.37124264,
        0.11074367, -0.47515884, -0.13795349, -0.33510417,  0.20475572,
        0.6537024 , -0.10462254,  0.28274846,  0.32527494, -0.14062619,
        0.31344542,  0.09137614, -0.16820167, -0.2747713 ,  0.03

6. Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных.
Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  vec += model_w2v[word].reshape((1, size))
и поделить финальный вектор на количество слов в твите.
На выходе должен получиться wordvec_df.shape = (49159, 200).

In [21]:
def vectorize(t, model):
    vector = np.zeros(200)
    norm = 0
    for w in t:
        if w in model:
            vector += model[w]
            norm += 1
    if norm > 0:
        vector = vector / norm
    return vector

In [23]:
combine_df['tweet_vector'] = combine_df['tweet_lemmatized'].apply(lambda x: vectorize(x, modelW2V))
combine_df.head(2)

id  ...                                       tweet_vector
0   1  ...  [-0.13146960043481418, -0.059886462188192775, ...
1   2  ...  [-0.08571832054294645, -0.12075696662068366, -...

[2 rows x 8 columns]

Чат-бот
Докрутить чат-бота (обучить лучший ft, w2v, взвесить idf). Модели сдампить и сложить в облако, ноутбук можно через гитхаб. Ответы: https://yadi.sk/d/SLgpFrmrBubeYg

# chat bot

In [31]:
import string
from pymorphy2 import MorphAnalyzer
from stop_words import get_stop_words
import annoy
from gensim.models import Word2Vec, FastText
import pickle
import numpy as np
from tqdm import tqdm_notebook

In [35]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [37]:
# assert False

#Small preprocess of the answers

question = None
written = False

with open("/content/prepared_answers.txt", "w") as fout:
    with open("/content/drive/My Drive/Otvety.txt", "r") as fin:
        for line in tqdm_notebook(fin):
            if line.startswith("---"):
                written = False
                continue
            if not written and question is not None:
                fout.write(question.replace("\t", " ").strip() + "\t" + line.replace("\t", " "))
                written = True
                question = None
                continue
            if not written:
                question = line.strip()
                continue

In [38]:
def preprocess_txt(line):
    spls = "".join(i for i in line.strip() if i not in exclude).split()
    spls = [morpher.parse(i.lower())[0].normal_form for i in spls]
    spls = [i for i in spls if i not in sw and i != ""]
    return spls

In [39]:
sentences = []

morpher = MorphAnalyzer()
sw = set(get_stop_words("ru"))
exclude = set(string.punctuation)
c = 0

with open("/content/drive/My Drive/Otvety.txt", "r") as fin:
    for line in tqdm_notebook(fin):
        spls = preprocess_txt(line)
        sentences.append(spls)
        c += 1
        if c > 100000:
            break

In [40]:
sentences = [i for i in sentences if len(i) > 2]

In [41]:
sentences[0]

['вопрос', 'тдв', 'отдыхать', 'лично', 'советовать', 'завести']

In [42]:
modelFT = FastText(sentences=sentences, size=200, min_count=2, window=5)

In [ ]:
tfidf_vectorizer = TfidfVectorizer(tokenizer=tokenizer,
                                   lowercase=False, max_df=0.9)

In [43]:
tfidf_vectorizer.fit_transform(sentences)


<64853x1000 sparse matrix of type '<class 'numpy.float64'>'
	with 355267 stored elements in Compressed Sparse Row format>

In [44]:
tfidf_vectorizer.idf_[tfidf_vectorizer.vocabulary_['вопрос']]

4.489546923041692

In [45]:
ft_index = annoy.AnnoyIndex(200 ,'angular')

index_map = {}
counter = 0

with open("/content/prepared_answers.txt", "r") as f:
    for line in tqdm_notebook(f):
        n_ft = 0
        spls = line.split("\t")
        index_map[counter] = spls[1]
        question = preprocess_txt(spls[0])
        
        vector_ft = np.zeros(200)
        for word in question:
            if word in modelFT:
                vector_ft += modelFT[word]
                n_ft += tfidf_vectorizer.idf_[tfidf_vectorizer.vocabulary_[word]] if word in tfidf_vectorizer.vocabulary_ else 1
        if n_ft > 0:
            vector_ft = vector_ft / n_ft
        ft_index.add_item(counter, vector_ft)
            
        counter += 1
        
        if counter > 100000:
            break

ft_index.build(50)

True

In [46]:
def get_response(question, index, model, index_map):
    question = preprocess_txt(question)
    vector = np.zeros(200)
    norm = 0
    for word in question:
        if word in model:
            vector += model[word]
            norm += tfidf_vectorizer.idf_[tfidf_vectorizer.vocabulary_[word]]
    if norm > 0:
        vector = vector / norm
    answers = index.get_nns_by_vector(vector, 3)
    return [index_map[i] for i in answers]

In [54]:
TEXT = "что такое счастье"

In [55]:
get_response(TEXT, ft_index, modelFT, index_map)

['Это когда у тя в жизни всё в поряде. \n',
 'Просыпаешся утром.Голова болит и настроение казалось бы х....Думаешь: "Че я здесь делаю, в этом городе, где нет возможностей и свободы, надо рвануть куда-н., быть может где-то меня ждет удача,,быть может я найду свою любовь...." Начинаешь вспоминать прожитые дни, прекрасные минуты жизни, друзей бесшабашных, хороших и чертовски красивых подруг, родных, и на душе становится просто КЛ.Идешь на кухню, открываешь холодильник и достаешь оттуда упаковку хорошего прохладного сока, выпиваешь и уже ничего не хочется.Забываешь про дурные мысли и продолжаешь жить.ВОТ ОНО СЧАСТЬЕ!!!. \n',
 'посмотри в зеркало, вот оно. \n']